### ISSUE TITLE
C++ exceptions are not properly shared between threads

### ISSUE THREAD

**Issue by arlebedev** [emscripten-core/emscripten#11503]
In the following minimal example:
```cpp
#include <emscripten.h>
#include <thread>

extern "C" int EMSCRIPTEN_KEEPALIVE
main()
{
    std::exception_ptr error;
    std::thread t([&error](){
        try {
            throw std::runtime_error("test");
        } catch (...) {
            error = std::current_exception();
        }
    });
    t.join();

    return 0;
}
```
```html
<!doctype html>
<html lang="en">
<head>
  <meta charset="utf-8">
  <title>WASM test</title>
  <script src="wasm_sample.js"></script>
</head>
<body>
    <script>
        CreateTest();
    </script>
</body>
</html>
```
```
em++ -O0 -g -s LLD_REPORT_UNDEFINED=1 -s ERROR_ON_UNDEFINED_SYMBOLS=0  -s MODULARIZE=1 -s DISABLE_EXCEPTION_CATCHING=0 -s USE_PTHREADS=1 -s EXPORT_NAME='CreateTest' -s PROXY_TO_PTHREAD=1  -std=c++1z -o main.cpp.o -c main.cpp
em++ -O0 -g -s LLD_REPORT_UNDEFINED=1 -s ERROR_ON_UNDEFINED_SYMBOLS=0 -s MODULARIZE=1 -s DISABLE_EXCEPTION_CATCHING=0 -s USE_PTHREADS=1 -s EXPORT_NAME='CreateTest' -s PROXY_TO_PTHREAD=1 main.cpp.o -o wasm_sample.js
```
the following error occurs:
```
wasm_sample.js:2303 Pthread 0x722ea0 exited.
threadExit @ wasm_sample.js:2303
onmessage @ wasm_sample.worker.js:136
wasm_sample.worker.js:24 worker.js onmessage() captured an uncaught exception: TypeError: Cannot read property 'refcount' of undefined
threadPrintErr @ wasm_sample.worker.js:24
onmessage @ wasm_sample.worker.js:170
wasm_sample.worker.js:24 TypeError: Cannot read property 'refcount' of undefined
    at ___exception_decRef (http://127.0.0.1:8000/cpp/build/wasm_sample.js:2649:19)
    at ___cxa_decrement_exception_refcount (http://127.0.0.1:8000/cpp/build/wasm_sample.js:2663:7)
    at std::exception_ptr::~exception_ptr() (wasm-function[125]:0x2fbc)
    at __original_main (wasm-function[56]:0x113e)
    at main (wasm-function[121]:0x2f82)
    at http://127.0.0.1:8000/cpp/build/wasm_sample.js:1993:22
    at ___call_main (http://127.0.0.1:8000/cpp/build/wasm_sample.js:2537:24)
    at __emscripten_thread_main (wasm-function[529]:0xcd88)
    at dynCall_ii (wasm-function[555]:0xd593)
    at Object.dynCall_ii (http://127.0.0.1:8000/cpp/build/wasm_sample.js:1993:22)
threadPrintErr @ wasm_sample.worker.js:24
onmessage @ wasm_sample.worker.js:171
wasm_sample.js:2438 pthread sent an error! http://127.0.0.1:8000/cpp/build/wasm_sample.worker.js:172: Uncaught TypeError: Cannot read property 'refcount' of undefined
worker.onerror @ wasm_sample.js:2438
error (async)
loadWasmModuleToWorker @ wasm_sample.js:2437
getNewWorker @ wasm_sample.js:2479
spawnThread @ wasm_sample.js:3563
_pthread_create @ wasm_sample.js:3767
proxy_main @ wasm_sample.wasm:1
(anonymous) @ wasm_sample.js:1993
callMain @ wasm_sample.js:4331
doRun @ wasm_sample.js:4374
run @ wasm_sample.js:4389
runCaller @ wasm_sample.js:4304
removeRunDependency @ wasm_sample.js:1894
receiveInstance @ wasm_sample.js:2064
receiveInstantiatedSource @ wasm_sample.js:2080
Promise.then (async)
(anonymous) @ wasm_sample.js:2103
Promise.then (async)
instantiateAsync @ wasm_sample.js:2101
createWasm @ wasm_sample.js:2128
(anonymous) @ wasm_sample.js:3880
(anonymous) @ (index):10
wasm_sample.worker.js:172 Uncaught TypeError: Cannot read property 'refcount' of undefined
    at ___exception_decRef (http://127.0.0.1:8000/cpp/build/wasm_sample.js:2649:19)
    at ___cxa_decrement_exception_refcount (http://127.0.0.1:8000/cpp/build/wasm_sample.js:2663:7)
    at std::exception_ptr::~exception_ptr() (wasm-function[125]:0x2fbc)
    at __original_main (wasm-function[56]:0x113e)
    at main (wasm-function[121]:0x2f82)
    at http://127.0.0.1:8000/cpp/build/wasm_sample.js:1993:22
    at ___call_main (http://127.0.0.1:8000/cpp/build/wasm_sample.js:2537:24)
    at __emscripten_thread_main (wasm-function[529]:0xcd88)
    at dynCall_ii (wasm-function[555]:0xd593)
    at Object.dynCall_ii (http://127.0.0.1:8000/cpp/build/wasm_sample.js:1993:22)
___exception_decRef @ wasm_sample.js:2649
___cxa_decrement_exception_refcount @ wasm_sample.js:2663
std::exception_ptr::~exception_ptr() @ 000a5e96:1
__original_main @ 000a5e96:1
main @ 000a5e96:1
(anonymous) @ wasm_sample.js:1993
___call_main @ wasm_sample.js:2537
__emscripten_thread_main @ 000a5e96:1
dynCall_ii @ 000a5e96:1
(anonymous) @ wasm_sample.js:1993
onmessage @ wasm_sample.worker.js:130
```

Seems there are exception info structures on JavaScript side which are thread-private. Similar errors occurs during any non-empty exception_ptr manipulation in threads other than the one, where the exception was created.
This bug makes it impossible to properly handle errors in real-world C++ projects. I heard that there is work in progress on a new EH mechanism but as I understand it is not yet so widely supported among all browsers so the legacy implementation also should be fixed. This is very critical for us so we are about to invest work into this issue (probably making our fork with the fix). We would like to contribute our effort into the mainline so the help from the core developers would be very appreciated, may be there are some ideas of the proper approach for the fix.

**Comment by arlebedev** [comment]
Currently my idea is to put all stuff from `___exception_infos` into the exception memory block itself. Looks pretty simple except `adjusted` field which is a list. I still cannot fully understand the logic behind it. Would be nice if somebody could explain this (@juj @kripken ?)
BTW in the example above execution goes into "shouldn't happen" code in `__cxa_find_matching_catch__deps`, this place:
```
    // Shouldn't happen unless we have bogus data in typeArray
    // or encounter a type for which emscripten doesn't have suitable
    // typeinfo defined. Best-efforts match just in case.
    thrown = {{{ makeGetValue('thrown', '0', '*') }}}; // undo indirection
    {{{ makeStructuralReturn(['thrown', 'throwntype']) }}};
```
Is that expected?

**Comment by kripken** [comment]
Yes, this is a known issue, see https://github.com/emscripten-core/emscripten/issues/11233 - those JS fields should be in C to fix this.

As mentioned in that issue, the new wasm exceptions support will not have this problem. But it would still be nice to fix this in the meantime if you are interested.

I don't understand all the logic in the JS exceptions code. But just mapping it to C should work in principle.

(Not sure what's going on in that last quoted code either, sorry - been a very long time since I worked on this.)

**Comment by arlebedev** [comment]
I tried to find the case where adjusted pointer is changed but did not find any. Obvious assumptions was that it is for pointer adjustment of exception class with multiple inheritance. But:
```cpp
#include <cstdio>
#include <exception>

class E1 {
public:
    int i;

    E1(int i): i(i)
    {}
};

class E2 {
public:
    int j;

    E2(int j): j(j)
    {}
};

class E3: public E1, E2 {
public:

    E3(int k): E1(k), E2(k + 1)
    {}
};

int
main()
{
    try {
        throw E3(42);
    } catch (int) {
        //XXX
    } catch (std::exception &) {
        printf("std::exception\n");
    // } catch (E1 &e) {
    //     printf("E1 %d\n", e.i);
    } catch (E2 &e) {
        printf("E2 %d\n", e.j);
    } catch (...) {
        printf("catch all\n");
    }

    return 0;
}
```
surprisingly this code does not catch E2, and falls into catch all block. E1 is caught if uncommented but no pointer conversion here. Checked it on [native compiler too](http://coliru.stacked-crooked.com/a/3ffd1edb8fd16983). Whole life is not enough to learn all C++ corner cases.
I am thinking about discarding this logic for the sake of simplicity (with some assertions about unchanged pointer).
BTW there are several suspicious places here.

https://github.com/emscripten-core/emscripten/blob/fa411e8fa9f28f38dbb1795e15ec9b0583004c44/src/library_exceptions.js#L125
Is this initial value needed here? Seems it is not.

https://github.com/emscripten-core/emscripten/blob/fa411e8fa9f28f38dbb1795e15ec9b0583004c44/src/library_exceptions.js#L300
Catch all block has type zero which is not handled here. Also handling exactly the same type can be optimized.

https://github.com/emscripten-core/emscripten/blob/fa411e8fa9f28f38dbb1795e15ec9b0583004c44/src/library_exceptions.js#L302
As I said I never seen the adjusted pointer is different. But in all cases it is added here, and with initial same value we have two base pointers in the list in most cases. Both seems to be useless.

https://github.com/emscripten-core/emscripten/blob/fa411e8fa9f28f38dbb1795e15ec9b0583004c44/src/library_exceptions.js#L309
This "shouldn't happen" block happens regularly. It is for "catch all", it is every frame without catch blocks (`typeArray` is empty in such case). Probably should be revised and at least comment changed.

**Comment by vagran** [comment]
I want to create pull request (I am switched to my private account) but I have no rights for pushing new branch. How ca I do this? I also can just attach path file here if necessary. Currently I have a working implementation which is accepted in our Emscripten fork we use in our project and it seems to be working well.

**Comment by vagran** [comment]
I have created PR https://github.com/emscripten-core/emscripten/pull/11518 (I have switched to my personal account). Probably some unit tests are needed as well. In our project we have adopted this patch in our fork and it works well. Enough for us. Hope it will be useful for someone else as well.

**Comment by vagran** [comment]
Seems it fails in asm.js which is unfortunately out of my scope. Hope someone will be able to fix it.

**Comment by kripken** [comment]
> Whole life is not enough to learn all C++ corner cases.

Very true...

I think the adjustment is for multiple virtual inheritance, which is indeed tricky. But we should have code in the test suite for it. Disabling just the adjustment and then running the suite should give some errors, and those tests can help understand the issue.

**Comment by vagran** [comment]
I have reproduced the adjusted pointer using exception with virtual inheritance (I also checked this case initially but somehow missed the adjustment in the debugger). Now question arises, how to get the initial pointer from the adjusted one. Having some shared thread-safe mapping looks too complicated for that. I also checked libcxxabi sources and they do not have this problem, all `cxa_*` functions expect the non-adjusted pointer or other type of entity (_Unwind_Exception) which is missing in JS (so general workflow is actually a bit different).
